Adding CountsStorageService to facilitate using the new counts store.

maxsumrall opened the pull request:
As the new counts store is still a work in progress, this is another necessary thing to have as part
 the current effort to start actually using the new counts store.

Commit:
Adding CountsStorageService to facilitate using the new counts store.

 As the new counts store is still a work in progress, this is another necessary thing to have as part
 the current effort to start actually using the new counts store.

Commit:
Write the snapshot to the transaction log if the snapshot is provided.

This is a very minor change, where the snapshot is deserialized to
the transaction log. The counts store is still not used,
nor is it every snapshotted, so a NO_SNAPSHOT field
is added to let current methods pass in a null snapshot
which will then not be written out.

Commit:
Add new counts store in recovery.

In recovery, this commit will have the new counts store initialized.

Commit:
Fix new counts store tests to deal with BASE_TX_ID being 1, not 0

Commit:
Apply updates to both counts stores

A new update applier is used to send updates to both of the counts stores.
Additional work still need to be done to get it working for batch import and migration.
 Some tests had to be updated to accommodate this change.

Commit:
Fix NPE due to a race with shutdown of the server

Commit:
Hook in the new counts store besides the old one

Commit:
Stop waiting for snapshot when DB is unhealthy

While waiting for a snapshot to have lower txid
transactions applied to it, we will now check if the
database health is not healthy, and quit waiting.
This should keep us from waiting forever while the
DB is shutting down, for example.

Also fixes a backup test which deleted the transaction logs
and still expected the DB to recover. This is not possible,
doubly so with the new counts store which snapshots to the
tx logs.

Commit:
Integrate new counts store in core-edge store copy situations.

Commit:
Mute HAClusterIT tests which no longer have valid assumptions.

The new counts store highlights issues with db files being deleted
and how we should now handle startup. Fixing that issues is a
separate task, which after completion this test should be unmuted.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    @Override
+    public Updater apply( long txId )
```
looks like a strange method that just delegate to another public method

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    @Override
+    public void accept( CountsVisitor visitor )
+    {
+        countsStore.forEach( ( countsKey, values ) -> {
```
traversal is not atomic here, you can see/not see actual values of one atomic state

maxsumrall replied:
Right. I've changed this to grab the write lock to ensure that the map doesn't change, but we are only using this in areas where I don't expect we are doing updates anyways, e.g. consistency checker.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+        return updaterFactory.getUpdater( countsStore, txId );
+    }
+
+    public IndexStatsUpdater indexStatsUpdater()
```
This is not used anywhere, so probably all IndexStatsUpdater can be deleted?

maxsumrall replied:
Yes, can be added again if we need it.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java:
```diff
@@ -62,6 +67,53 @@ public InMemoryCountsStore()
+            map.compute( key, ( k, v ) -> {
+                if ( v == null )
+                {
+                    return Arrays.copyOf( delta, delta.length );
```
do we really need to copy always incoming array? that will create tons of array copies, not only here but on all places where we copy and apply updates?
Maybe final data structure to hold values?

maxsumrall replied:
I've add this work to the counts store trello card to be addressed later.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java:
```diff
@@ -156,6 +209,12 @@ public CountsSnapshot snapshot( long txId )
+    @Override
+    public void forEach( BiConsumer<CountsKey,long[]> action )
+    {
+        map.forEach( action );
```
atomicity is broken here

maxsumrall replied:
Same answer to the problem with accept() in CountsStorageService.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/ReadOnlyCountsStorageService.java:
```diff
@@ -0,0 +1,96 @@
+import org.neo4j.kernel.internal.DatabaseHealth;
+import org.neo4j.register.Register;
+
+public class ReadOnlyCountsStorageService implements CountsStorageService
```
not used at all

maxsumrall replied:
Yes, removed for now.

davidegrohmann replied:
No, I think we should make sure we use it when we start the db readonly.

maxsumrall replied:
We can put it back when we start using it.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/RecordStorageEngine.java:
```diff
@@ -399,6 +404,17 @@ public void init() throws Throwable
         labelScanStore.init();
     }
 
+    public void initFromSnapshot(CountsSnapshot snapshot)
```
i do not really think that RecordStorageEngine is a good place for those 2 methods

davidegrohmann replied:
Which is a good place for those methods?

MishaDemianenko replied:
lets talk tomorrow about that :)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/LogRecoveryInfo.java:
```diff
@@ -17,30 +17,30 @@
 {
-    final private Map<CountsKey,long[]> map;
-    final private long txId;
+    LogPosition logPosition;
```
private fields?

maxsumrall replied:
Yes.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/recovery/PositionToRecoverFrom.java:
```diff
@@ -47,18 +49,20 @@ public PositionToRecoverFrom( LatestCheckPointFinder checkPointFinder )
      */
-    @Override
-    public LogPosition apply( long currentLogVersion ) throws IOException
+    public LogRecoveryInfo apply( long currentLogVersion ) throws IOException
```
should we just return CheckPoint object here instead? it sounds reasonable that we have a checkpoint to be able to do recovery from them, so why not to just use them? wdyt?

maxsumrall replied:
We are also returning the log recovery position which is not related to the checkpoint, so LogRecoveryInfo seems to make more sense at the moment.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStoreIntegrationTest.java:
```diff
@@ -178,7 +178,7 @@ public void run()
             {
                 int id = intermediateStateTestManager.getId();
-                long txId = id + ThreadLocalRandom.current().nextLong( 0, 5 );
+                long txId = id; //+ ThreadLocalRandom.current().nextLong( 0, 5 );
```
hm...:flashlight: looks like should be uncommented?

maxsumrall replied:
👻 Yes.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/DualCountsStoreUpdater.java:
```diff
@@ -0,0 +1,72 @@
+    {
+        countsStorageUpdater = countsStorageService.apply( txId );
+        Optional<CountsAccessor.Updater> result = countsTracker.apply( txId );
+        this.countsTrackerUpdater = result.isPresent() ? result.get() : null;
```
`result.orElse(null)`

lutovich replied:
Mix of `Optional` and `null` looks quite strange. Can `CountsStorageService#apply()` return nullable `CountsAccessor.Updater` instead?

lutovich replied:
Or another idea: `CountsStorageService#apply()` always returns non-null. When real updater can't be created some empty impl is returned. With this approach no null checks would be needed.

maxsumrall replied:
It is the updater from the old counts store, CountsTracker, that might return null.

MishaDemianenko commented on enterprise/backup/src/main/java/org/neo4j/backup/OnlineBackupKernelExtension.java:
```diff
@@ -70,17 +70,17 @@
     private final LogProvider logProvider;
     private final Monitors monitors;
-    private BackupServer server;
+    private volatile BackupServer server;
```
Can you please describe the case when you need this field to be volatile?

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/DatabasePanicEventGenerator.java:
```diff
@@ -36,15 +33,6 @@ public DatabasePanicEventGenerator( KernelEventHandlers kernelEventHandlers )
-            }
-        } );
-        executor.shutdown();
+        kernelEventHandlers.kernelPanic( error, cause ) ;
```
Why was this change made? Is it relevant to the PR?

maxsumrall replied:
Was just some funny code.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsSnapshot.java:
```diff
@@ -0,0 +1,94 @@
+
+    public CountsSnapshot( long txId )
+    {
+        this( txId, new ConcurrentHashMap<>() );
```
All other maps in this class are not concurrent. Why this one is? Imo if this class should be thread-safe than `Map` should be `ConcurrentMap` everywhere.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    public void initialize( CountsSnapshot snapshot, DatabaseHealth databaseHealth )
+    {
+        countsStore = new InMemoryCountsStore( snapshot, databaseHealth );
+        indexStatsUpdater = new IndexStatsUpdaterFactory().indexStatsUpdater( countsStore );
```
Can `IndexStatsUpdaterFactory` be removed? Feels like we can create `indexStatsUpdater` directly.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/keys/JustASimpleChannel.java:
```diff
@@ -0,0 +1,107 @@
+
+import org.neo4j.kernel.impl.transaction.log.FlushableChannel;
+
+public class JustASimpleChannel implements FlushableChannel
```
Looks like this class is not used and should be removed.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/StoreMigrator.java:
```diff
@@ -693,7 +706,12 @@ public void rebuildCounts( File storeDir, String versionToMigrateFrom ) throws I
+            long lastCommittedTx = MetaDataStore.getRecord( pageCache, neoStore, Position.LAST_TRANSACTION_ID );
+
+            // write a check point in the log in order to make recovery work in the newer version
+            new StoreMigratorCheckPointer( storeDir, fileSystem ).checkPoint( logVersion, lastCommittedTx, snapshot );
```
Probably better to extract method for this line and 688.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointInfo.java:
```diff
@@ -0,0 +1,28 @@
+
+import org.neo4j.kernel.impl.store.counts.CountsSnapshot;
+
+public interface CheckPointInfo
```
Why this is not a simple class with two fields?

lutovich commented on community/kernel/src/main/java/org/neo4j/storageengine/api/StorageEngine.java:
```diff
@@ -113,6 +114,14 @@ void createCommands(
+     * As part of the simplified counts store. This will initialize the count store in the storage engine
+     * when doing recovery.
+     */
+    void initFromSnapshot( CountsSnapshot snapshot );
```
Feels strange to have methods related to counts store exposed in storage engine API. Shouldn't it be unaware of things that are underneath it?

davidegrohmann replied:
Where to add that instead? Note that a snapshot is pretty general (a long and a map) and unrelated to storage implementation details.

lutovich replied:
Unfortunately I can't think of a better place right now.
These are other things that concern me:
- counts store initialization grew through many layers, would be good to keep it more isolated
- we expose not only counts store from `StorageEngine` but also the fact that it is initialized from the outside

Would be good to discuss this thing in person.

lutovich commented on enterprise/ha/src/test/java/org/neo4j/ha/TestPullUpdatesApplied.java:
```diff
@@ -152,7 +153,9 @@ public void failed( InstanceId server )
         restart( toKill ); // recovery and branching.
-        boolean hasBranchedData = new File( targetDirectory, "branched" ).listFiles().length > 0;
+        File[] branchedFiles = new File( targetDirectory, "branched" ).listFiles();
+        assertNotNull( branchedFiles );
```
`branchedFiles` could be null when `File( targetDirectory, "branched" )` does not exist. Thus `assertNotNull()` looks a bit strange.

davidegrohmann replied:
If it was `null` and we would have blown up on the `.length` with a NPE. Isn't it better having an assertion `!= null`?

lutovich replied:
I understand the NPE possibility but `branchedFiles == null` feels like a completely valid case for this test. This is not super important however.

Commit:
Various fixes and updates based on feedback

-No longer applied changes in recovery from transactions already applied
to the counts store.
-Other various code-cleanup changes based on feedback.
